compressors/shannon_coder.py

```diff
@@ -0,0 +1,156 @@
+"""Shannon Coding
+Shannon coding is a basic prefix free code. There is some confusion in literature about Shannon, Fano and
+Shannon-Fano codes, e.g. see Naming section of https://en.wikipedia.org/wiki/Shannon–Fano_coding.
+
+This document uses cumulative probability based method for shannon coding as described in the wiki article above.
+"""
+from typing import Any, Tuple
+from utils.bitarray_utils import float_to_bitarrays, BitArray
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from compressors.prefix_free_compressors import PrefixFreeEncoder, PrefixFreeDecoder, PrefixFreeTree, BinaryNode
+from core.prob_dist import ProbabilityDist
+import math
+
+
+class ShannonTree(PrefixFreeTree):
+    """
+    Generates Codewords based on Shannon Coding. Allows generating PrefixFreeTree from the obtained codewords.
+    """
+
+    def __init__(self, prob_dist: ProbabilityDist):
+        self.prob_dist = prob_dist
+        # sort the probability distribution in decreasing probability and get cumulative probability which will be
+        # used for encoding
+        self.sorted_prob_dist = ProbabilityDist.get_sorted_prob_dist(prob_dist.prob_dict)
+        self.cum_prob_dict = self.sorted_prob_dist.cumulative_prob_dict
+        # construct the tree and set the root_node of PrefixFreeTree base class
+        super().__init__(root_node=self.build_shannon_tree())
+
+    def encode_alphabet(self, symbol) -> BitArray:
```

Comment:
why not just call it `encode_symbol` maybe with underscore if we want to distinguish from the prefix coder version?

Reply:
it's encoding an alphabet in the dictionary not a symbol. I wanted to be clear wrt alphabet and symbol. It will only be used once when the original alphabet is defined. I have modified it to` _encode_alphabet`. Lmk if u have any other concerns.

Reply:
Also move the whole function of generating tree from codewords to PrefixFreeCompressor